numpy/lib/index_tricks.py

```diff
@@ -596,6 +597,419 @@ def __iter__(self):
     next = __next__
 
 
+class ndrange(_nx.collections_abc.Sequence):
+    """
+    The ndrange type represents an immutable sequence of numbers and can be
+    used for looping over specific multi-dimensional indicies in ``for`` loops.
+
+    # TODO: how do I get this to show up in the docs
+    class ndrange(stop)
+    class ndrange(start, stop[, step])
```

Comment:
This should happen automatically. It doesn't?

Reply:
This is refering to the craziness where if only a single argument is passed in, it is considered `stop`, otherwise the first argument is considered `start`.  I think Ij ust have to put it as the very first line.

Reply:
Yes, that's the behavior of `range` too. 

I don't think you need to quote the signature anywhere though, python/sphinx prints it out automatically where needed, eg at https://docs.scipy.org/doc/numpy-1.15.1/reference/routines.indexing.html

Reply:
To clarify, you shouldn't have these signature lines here.

The very first line of a docstring should be a shorthand description of the class, since other parts of numpy will pull in just that first line in overview documentation pages.

Reply:
But then where do you explain what the `first positional parameter means`. It can mean `start` or `stop` depending on the number of positional arguments.

Reply:
Hmm yeah, that is some special behavior which should be described. But in any case we can't put that as the first line because of how the numpy documentation system works.

I would add a sentence or two about it at the end of the "parameters", something like "When called with a signle positional argument that argument is used as `stop` and start is assumed to be 0. When called with two or three positional arguments....". Then maybe as the second paragraph of the docstring also say `Can be called as "ndrange(stop)" or "ndrange(start, stop, [step])", see parameters below.`

Reply:
OK. reworeded. range in python is even weirder:
```
>>> range(start=0, stop=4, step=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: range() does not take keyword arguments
```

I think ndrange, should definitely take keyword arguments. becaues having multiple tuples can become confusing.

Even `slice` doesn't take keyword arguments.
```
>>> slice(start=4, end=8, step=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: slice() does not take keyword arguments
>>> slice(4, 8).start
4
```